File: src/sort/impl/QuickSort.java
package sort.impl;

import sort.AbstractSort;

public class QuickSort extends AbstractSort {
    public QuickSort(int[] input) {
        super(input);
    }

    public int[] sort() {
        quickSort(arr, 0, arr.length - 1);
        return arr;
    }

    private static void quickSort(int[] arr, int start, int end) {
        int mid = getMid(arr, start, end);

        if (mid > start)
            quickSort(arr, start, mid - 1);
        if (mid < end)
            quickSort(arr, mid + 1, end);
    }

    private  static int getMid(int[] arr, int left, int right) {
        int flag = arr[left];

        while(left < right) {
            while(left < right && arr[right] >= flag)
                right--;
            arr[left] = arr[right];
            while(left < right && arr[left] <= flag)
                left++;
            arr[right] = arr[left];
        }

        arr[left] = flag;
        return left;
    }
}

File: python/sort_methods/quick_sort.py
"""
快速排序
2021/10/12
"""
def qsorted(arr):
    __do_qsort(arr, 0, len(arr) - 1)


def __do_qsort(arr, start, end):
    if end - start < 1:
        return
    
    flag = arr[start]
    h = end
    l = start
    while l < h:
        while arr[h] >= flag and l < h:
            h -= 1
        arr[l] = arr[h]
        while arr[l] <= flag and l < h:
            l += 1
        arr[h] = arr[l]
    
    arr[l] = flag
    __do_qsort(arr, start, l - 1)
    __do_qsort(arr, l + 1, end)


if __name__ == "__main__":
    tlist = [7, 2, 15, 11]
    qsorted(tlist)
    print(tlist)
